pkg/querier/tail.go

```diff
@@ -1,92 +1,334 @@
 package querier
 
 import (
-	"context"
+	"fmt"
+	"sync"
 	"time"
 
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
 	"github.com/grafana/loki/pkg/iter"
 	"github.com/grafana/loki/pkg/logproto"
+	"github.com/pkg/errors"
 )
 
-// tailIteratorIncrement is for defining size of time window we want to query entries for
-const tailIteratorIncrement = 10 * time.Second
+const (
+	// if we are not seeing any response from ingester, how long do we want to wait by going into sleep
+	nextEntryWait = time.Second / 2
 
-// delayQuerying is for delaying querying of logs for specified seconds to not miss any late entries
-const delayQuerying = 10 * time.Second
+	// keep checking connections with ingesters in duration
+	checkConnectionsWithIngestersPeriod = time.Second * 5
 
-func (q *Querier) tailQuery(ctx context.Context, queryRequest *logproto.QueryRequest) iter.EntryIterator {
-	queryRequest.Start = time.Now().Add(-(tailIteratorIncrement + delayQuerying))
-	queryRequest.Direction = logproto.FORWARD
+	bufferSizeForTailResponse = 10
+)
+
+type droppedEntry struct {
+	Timestamp time.Time
+	Labels    string
+}
+
+// TailResponse holds response sent by tailer
+type TailResponse struct {
+	Streams        []logproto.Stream `json:"streams"`
+	DroppedEntries []droppedEntry    `json:"dropped_entries"`
```

Comment:
Can you briefly explain what are dropped entries ? What are they for ?

Reply:
If the tail client is slow, we start dropping the entries and store its details here so that client knows we dropped some entries because its slow and can query for it if needed.